src/domain/core/src/utils/time_source.rs

```diff
@@ -24,46 +30,155 @@ pub trait SystemTimeSource: Send + Sync {
 #[dill::interface(dyn SystemTimeSource)]
 pub struct SystemTimeSourceDefault;
 
+#[async_trait::async_trait]
 impl SystemTimeSource for SystemTimeSourceDefault {
     fn now(&self) -> DateTime<Utc> {
         Utc::now()
     }
+
+    async fn sleep(&self, duration: Duration) {
+        let std_duration = duration.to_std().unwrap();
+
+        tokio::time::sleep(std_duration).await;
+    }
 }
 
 /////////////////////////////////////////////////////////////////////////////////////////
 
+#[derive(Debug)]
+struct AwaitingCaller {
+    wake_up_time: DateTime<Utc>,
+    waker_tx: Option<oneshot::Sender<()>>,
+}
+
+impl Eq for AwaitingCaller {}
+
+impl PartialEq<Self> for AwaitingCaller {
+    fn eq(&self, other: &Self) -> bool {
+        self.cmp(other) == Ordering::Equal
+    }
+}
+
+impl PartialOrd<Self> for AwaitingCaller {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for AwaitingCaller {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.wake_up_time.cmp(&other.wake_up_time)
+    }
+}
```

Comment:
Isn't auto-derive comparison doing the same?

Reply:
Unfortunately without additional crates we can't do this